### ISSUE TITLE
Partition column gets dropped from dataframe when reading parquet file list using pyarrow engine

### ISSUE THREAD

**Issue by senglehardt** [dask/dask#7600]
**What happened**:
When using dask to read a list of parquet files in a dataset with hive partitioning using the `engine="pyarrow"`, the partitioning column(s) gets dropped. In this case the parquet files were written using pyspark.
![image](https://user-images.githubusercontent.com/30219769/115909022-89aaf500-a430-11eb-8f07-097ba29f4cfa.png)

I have not been able to confirm the same behavior using `engine="pyarrow-dataset"` because it results in another error as posted here [post on stackoverflow](https://stackoverflow.com/questions/67222212/read-partitioned-parquet-dataset-written-by-spark-using-dask-and-pyarrow-dataset),
which was an already-known bug here #7512 .

`engine="pyarrow-dataset"` is ultimately what I hope to use in case this issue spills over to both engines.

**What you expected to happen**:
The partitioning column should be a column returned in the dask dataframe.

**Minimal Complete Verifiable Example**:

```python

import os
from glob import glob
import pyarrow.parquet as pq
import pyarrow.dataset as ds
import dask.dataframe as dd
from pyspark.sql import SparkSession

home_dir = "/home/xxx/" # replace with your base directory

spark = SparkSession.builder.master("local[*]").getOrCreate()
df = spark.createDataFrame([("A", 1, 4), ("B", 2, 5), ("C", 3, 6)],["id", "col1", "col2"])
df.write.format("parquet").mode("overwrite").partitionBy("id").save(home_dir + "parquet_dataset")
paths = [y for x in os.walk(home_dir + "parquet_dataset") for y in glob(os.path.join(x[0],'*.parquet'))]
dd.read_parquet(path=paths, engine="pyarrow").compute()

```

But the following pyarrow methods work fine:
```
pq.read_table(paths).to_pandas()
ds.dataset(source=paths, partitioning="hive", format="parquet").to_table().to_pandas()
```
![image](https://user-images.githubusercontent.com/30219769/115909216-d68ecb80-a430-11eb-8998-4745fb06d690.png)

**Anything else we need to know?**:
Let me know if I missed providing anything else useful. Thank you for taking a look.

**Environment**:

- Dask version: 2021.4.0
- Python version: 3.8
- Operating System: Ubuntu
- Install method (conda, pip, source): pip
- pyarrow version: 3.0.0
- pandas version: 1.2.4

**Comment by rjzamora** [comment]
Thanks for raising @senglehardt - Can you check that #7541 doesn't address this issue?  If not, I'll be happy to try to track down a fix.  However, I have no experience with pyspark, so it may be a bit faster if I have a pandas-based reproducer :)

**Comment by senglehardt** [comment]
Thank you @rjzamora.

Regarding issue #7512, according to this [SO](https://stackoverflow.com/questions/56743868/drop-partition-columns-when-writing-parquet-in-pyspark) post, spark does not write the partitionBy columns to the parquet file. I came across this while searching for a way to not have Spark write the partitions to columns in the parquet file in order to circumvent the dask error using the `pyarrow-dataset` engine. Turns out, spark does not do this assuming we take the first answer at face value (although I have not checked with parquet tools).

#7541 considers when the partition is also written to the column so I am not sure if it addresses this. I think #this (#7600) has to do with passing a list of files. Only the version providing a list of paths drops the column using the `ArrowLegacyEngine`:
`dd.read_parquet(path=home_dir + "parquet_dataset", engine="pyarrow").compute()` # Works
`dd.read_parquet(path=paths, engine="pyarrow").compute()` # Drops the column

Now that I think about it, I am not sure how `pyarrow.dataset` is able to read the list of file paths unless it is extracting the partition from the filename or if this is included in the metadata of the parquet file. 

I'll follow up in a little while to see if I can replicate this using a pandas-based reproducer. Thanks again!

**Comment by senglehardt** [comment]
Interestingly, writing the parquet using pandas fixed the `engine=pyarrow-dataset` issue but not the `ArrowLegacyEngine` issue for reading in a list of files:
```
import pandas as pd
import dask.dataframe as dd
from glob import glob

data = {"id": ["A", "B", "C"],
        "col1": [1, 2, 3],
        "col2": [4, 5, 6]}
pdf = pd.DataFrame(data)
pdf.to_parquet(path=home_dir+"parquet_dataset2", engine="pyarrow", compression="snappy", partition_cols="id")

paths = [y for x in os.walk(home_dir + "parquet_dataset2") for y in glob(os.path.join(x[0], '*.parquet'))]

dd.read_parquet(path=home_dir + "parquet_dataset2", engine="pyarrow").compute() # Still works
dd.read_parquet(path=paths, engine="pyarrow").compute() # Still does not work

dd.read_parquet(path=home_dir + "parquet_dataset2", engine="pyarrow-dataset").compute() # NOW WORKS!
dd.read_parquet(path=paths, engine="pyarrow-dataset").compute() # NOW WORKS!
```

So once #7512 is merged into new dask release, I should be golden (as I need to read spark written partitioned parquet file list).

**Comment by rjzamora** [comment]
>Now that I think about it, I am not sure how pyarrow.dataset is able to read the list of file paths unless it is extracting the partition from the filename or if this is included in the metadata of the parquet file.

Yes - pyarrow.dataset is extracting partitioning information from the path when it processes either a list of files or when it needs to walk through the files within a root directory.  By default, pyarrow.dataset assumes that you may have "hive"-like partitioning.  

The legacy API does NOT make any assumptions about dataset partitioning if you pass in a list of files.  It will only extract partition information from the paths if you pass in a root directory (so that pyarrow is forced to walk through the files itself).  For now, it seems that the only way to read a partitioned dataset "correctly" with the legacy API is to pass in the root directory, OR pass in a list of files that includes a "_metadata" file (although I still need to confirm the latter case myself).

**Comment by jrbourbeau** [comment]
@senglehardt just checking in here, are you still running into issues or does upgrading to a newer version of `dask` resolve the issue?

**Comment by jrbourbeau** [comment]
Closing due to lack of activity. Feel free to re-open as needed
